<a href="https://colab.research.google.com/github/regitaaft/iris-eda-ml/blob/main/UTS_ML.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

# Load data
train = pd.read_csv('data_training.csv')
test = pd.read_csv('data_testing.csv')

Library pandas digunakan untuk membaca dan mengolah dataset. Selanjutnya, fungsi train_test_split digunakan untuk membagi data menjadi data pelatihan dan validasi. Model yang digunakan dalam penelitian ini adalah Random Forest Classifier karena memiliki performa yang baik pada data klasifikasi. Selain itu, digunakan fungsi accuracy_score untuk mengevaluasi tingkat akurasi model. Dataset training dan testing kemudian dibaca menggunakan fungsi read_csv().

In [ ]:
train = pd.read_csv('data_training.csv')
test = pd.read_csv('data_testing.csv')
train.head()

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality,Id
0,7.3,0.740,0.08,1.7,0.094,10.0,45.0,0.99576,3.24,0.50,9.8,5,1366
1,8.1,0.575,0.22,2.1,0.077,12.0,65.0,0.99670,3.29,0.51,9.2,5,103
2,10.1,0.430,0.40,2.6,0.092,13.0,52.0,0.99834,3.22,0.64,10.0,7,942
3,12.9,0.500,0.55,2.8,0.072,7.0,24.0,1.00012,3.09,0.68,10.9,6,811
4,8.4,0.360,0.32,2.2,0.081,32.0,79.0,0.99640,3.30,0.72,11.0,6,918


Output menampilkan lima baris pertama dari dataset training yang berisi variabel-variabel kimiawi anggur seperti acidity, chlorides, alcohol, dan variabel target quality.

In [ ]:
train.info()
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 857 entries, 0 to 856
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         857 non-null    float64
 1   volatile acidity      857 non-null    float64
 2   citric acid           857 non-null    float64
 3   residual sugar        857 non-null    float64
 4   chlorides             857 non-null    float64
 5   free sulfur dioxide   857 non-null    float64
 6   total sulfur dioxide  857 non-null    float64
 7   density               857 non-null    float64
 8   pH                    857 non-null    float64
 9   sulphates             857 non-null    float64
 10  alcohol               857 non-null    float64
 11  quality               857 non-null    int64  
 12  Id                    857 non-null    int64  
dtypes: float64(11), int64(2)
memory usage: 87.2 KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 286 entries, 0 to 285
Data c

Fungsi info() digunakan untuk melihat struktur dataset, tipe data, dan jumlah data non-null pada setiap variabel. Hasil output menunjukkan bahwa dataset tidak memiliki missing value dan seluruh variabel bertipe numerik sehingga dapat langsung digunakan untuk analisis lebih lanjut.

In [ ]:
train.isnull().sum()
test.isnull().sum()

,0
fixed acidity,0
volatile acidity,0
citric acid,0
residual sugar,0
chlorides,0
free sulfur dioxide,0
total sulfur dioxide,0
density,0
pH,0
sulphates,0


Fungsi isnull().sum() digunakan untuk mengecek jumlah missing value pada setiap variabel. Hasil menunjukkan bahwa seluruh variabel memiliki nilai missing value sebesar 0 sehingga data siap digunakan untuk proses analisis dan pemodelan.

In [ ]:
X = train.drop('quality', axis=1)
y = train['quality']

print(X.shape)
print(y.shape)

(857, 12)
(857,)


Berdasarkan output, dataset memiliki 857 data observasi dengan 12 variabel fitur yang digunakan sebagai input model. Selain itu, terdapat 857 data target quality yang akan diprediksi oleh model klasifikasi.

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_val, y_train, y_val = train_test_split(
    X, y, test_size=0.2, random_state=42
)

print(X_train.shape)
print(X_val.shape)

(685, 12)
(172, 12)


Dataset dibagi menjadi data training dan data testing dengan proporsi 80:20 menggunakan fungsi train_test_split(). Data pelatihan digunakan untuk membangun model machine learning, sedangkan data validasi digunakan untuk mengevaluasi performa model pada data yang belum pernah dilihat sebelumnya.

In [ ]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train = scaler.fit_transform(X_train)
X_val = scaler.transform(X_val)
test_scaled = scaler.transform(test)

Standardisasi data dilakukan menggunakan StandardScaler untuk menyamakan skala antar variabel. Proses ini membantu model machine learning bekerja lebih optimal karena setiap fitur memiliki rentang nilai yang berbeda.

## MODEL 1 Random Forest

In [ ]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

rf = RandomForestClassifier(random_state=42)
rf.fit(X_train, y_train)

y_pred_rf = rf.predict(X_val)

print("Accuracy RF:", accuracy_score(y_val, y_pred_rf))

Accuracy RF: 0.5988372093023255


Model Random Forest digunakan untuk melakukan klasifikasi kualitas anggur. Berdasarkan hasil evaluasi, model menghasilkan nilai akurasi sebesar 59,88%, yang menunjukkan bahwa model mampu memprediksi sebagian besar data dengan cukup baik.

## Confussion Matriks

In [ ]:
from sklearn.metrics import confusion_matrix

cm = confusion_matrix(y_val, y_pred_rf)
print(cm)

[[ 0  1  2  0  0]
 [ 0 47 20  0  0]
 [ 0 23 47  8  0]
 [ 0  1 11  9  0]
 [ 0  0  0  3  0]]


Confusion matrix digunakan untuk mengevaluasi hasil klasifikasi model Random Forest.
Model cukup baik dalam memprediksi kelas tengah, terutama pada kelas dengan jumlah data lebih banyak.
Terdapat 47 data pada salah satu kelas yang berhasil diprediksi dengan benar.
Namun, model masih mengalami kesalahan klasifikasi pada beberapa kelas yang berdekatan, misalnya antara kualitas 5 dan 6.
Selain itu, terdapat beberapa kelas dengan jumlah data sedikit yang belum dapat diprediksi dengan baik oleh model.
Hal ini menunjukkan bahwa distribusi data yang tidak seimbang mempengaruhi performa model dalam memprediksi seluruh kelas secara merata.

## Prediksi Data Set

In [ ]:
test_pred = rf.predict(test_scaled)

print(test_pred[:10])

[5 5 5 5 5 6 5 5 6 5]


Output menunjukkan hasil prediksi kualitas anggur pada 10 data pertama di dataset testing.

Hasil prediksi didominasi oleh kelas 5 dan 6 karena kedua kelas tersebut merupakan kelas yang paling banyak muncul pada dataset training. Hal ini menunjukkan bahwa model cenderung memprediksi kelas yang memiliki jumlah data dominan.

In [ ]:
output = pd.DataFrame({
    'id': range(len(test_pred)),
    'quality': test_pred
})

output.to_csv('hasilprediksi_02.csv', index=False)

output.head()

,id,quality
0,0,5
1,1,5
2,2,5
3,3,5
4,4,5


## Kesimpulan
Berdasarkan hasil analisis dan pemodelan machine learning pada dataset Wine Quality, dapat disimpulkan bahwa algoritma Random Forest berhasil digunakan untuk melakukan klasifikasi kualitas anggur berdasarkan variabel-variabel kimiawi yang tersedia pada dataset.

Sebelum proses pemodelan dilakukan, data terlebih dahulu melalui tahap persiapan dan pengecekan missing value. Hasil pengecekan menunjukkan bahwa dataset training maupun testing tidak memiliki data yang hilang sehingga data dapat langsung digunakan pada proses analisis. Selanjutnya, data dibagi menjadi data pelatihan dan data validasi dengan proporsi 80:20 untuk mengevaluasi performa model.

Berdasarkan hasil evaluasi menggunakan metrik accuracy, model Random Forest menghasilkan nilai akurasi sebesar sekitar 59,88%. Hasil confusion matrix menunjukkan bahwa model cukup baik dalam memprediksi kelas kualitas yang dominan, terutama pada kelas tengah, namun masih mengalami kesalahan klasifikasi pada beberapa kelas yang memiliki karakteristik mirip dan jumlah data yang lebih sedikit. Hal ini menunjukkan bahwa distribusi data yang tidak seimbang mempengaruhi performa model dalam memprediksi seluruh kelas secara merata.

Setelah model selesai dilatih dan dievaluasi, model digunakan untuk memprediksi kualitas anggur pada dataset testing.